File: resources/expected/ros2/python-isolette/strict/src/isolette_py_pkg/isolette_py_pkg/base_code/temperature_sensor_cpi_thermostat_base_src.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from collections import deque
from typing import Union
import threading
from rclpy.callback_groups import ReentrantCallbackGroup
from isolette_py_pkg_interfaces.msg import PhysicalTempimpl
from isolette_py_pkg_interfaces.msg import TempWstatusimpl

#========================================================
# Re-running Codegen will overwrite changes to this file
#========================================================

class temperature_sensor_cpi_thermostat_base(Node):
    def __init__(self):
        super().__init__("temperature_sensor_cpi_thermostat")

        self.cb_group_ = ReentrantCallbackGroup()

        self.lock_ = threading.Lock()

        # Setting up connections
        self.temperature_sensor_cpi_thermostat_air_subscription_ = self.create_subscription(
            PhysicalTempimpl,
            "temperature_sensor_cpi_thermostat_air",
            self.accept_air,
            1,
            callback_group=self.cb_group_)

        self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_1 = self.create_publisher(
            TempWstatusimpl,
            "thermostat_monitor_temperature_manage_monitor_interface_mmit_current_tempWstatus",
            1)

        self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_2 = self.create_publisher(
            TempWstatusimpl,
            "thermostat_monitor_temperature_manage_alarm_mat_current_tempWstatus",
            1)

        self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_3 = self.create_publisher(
            TempWstatusimpl,
            "thermostat_monitor_temperature_manage_monitor_mode_mmmt_current_tempWstatus",
            1)

        self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_4 = self.create_publisher(
            TempWstatusimpl,
            "thermostat_regulate_temperature_manage_regulator_interface_mrit_current_tempWstatus",
            1)

        self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_5 = self.create_publisher(
            TempWstatusimpl,
            "thermostat_regulate_temperature_manage_heat_source_mhst_current_tempWstatus",
            1)

        self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_6 = self.create_publisher(
            TempWstatusimpl,
            "thermostat_regulate_temperature_manage_regulator_mode_mrmt_current_tempWstatus",
            1)

        # timeTriggeredCaller callback timer
        self.periodTimer_ = self.create_timer(1, self.timeTriggeredCaller, callback_group=self.cb_group_)

        self.infrastructureIn_air = deque()
        self.applicationIn_air = deque()

        self.infrastructureOut_current_tempWstatus = deque()
        self.applicationOut_current_tempWstatus = deque()

        # Used by receiveInputs
        self.inDataPortTupleVector = [
            [self.infrastructureIn_air, self.applicationIn_air]
        ]

        # Used by receiveInputs
        self.inEventPortTupleVector = [
            [self.infrastructureIn_air, self.applicationIn_air]
        ]

        # Used by sendOutputs
        self.outPortTupleVector = [
            [self.applicationOut_current_tempWstatus, self.infrastructureOut_current_tempWstatus, self.sendOut_current_tempWstatus]
        ]

    def init_air(self, val):
        self.enqueue(self.infrastructureIn_air, val)


    def timeTriggered(self):
        raise NotImplementedError("Subclasses must implement this method")

    #=================================================
    #  C o m m u n i c a t i o n
    #=================================================

    def accept_air(self, msg):
        self.enqueue(self.infrastructureIn_air, msg)

    def get_air(self):
        msg = self.applicationIn_air[0]
        return msg

    def sendOut_current_tempWstatus(self, msg):
        if type(msg) is TempWstatusimpl:
            self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_1.publish(msg)
            self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_2.publish(msg)
            self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_3.publish(msg)
            self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_4.publish(msg)
            self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_5.publish(msg)
            self.temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_6.publish(msg)
        else:
            self.get_logger().error("Sending out wrong type of variable on port current_tempWstatus.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.")

    def put_current_tempWstatus(self, msg):
        self.enqueue(self.applicationOut_current_tempWstatus, msg)

    def timeTriggeredCaller(self):
        self.receiveInputs()
        self.timeTriggered()
        self.sendOutputs()

    def receiveInputs(self):
        for port in self.inDataPortTupleVector:
            infrastructureQueue = port[0]
            if not(len(infrastructureQueue) == 0):
                msg = infrastructureQueue[0]
                self.enqueue(port[1], msg)

        for port in self.inEventPortTupleVector:
            infrastructureQueue = port[0]
            if not(len(infrastructureQueue) == 0):
                msg = infrastructureQueue[0]
                infrastructureQueue.pop()
                self.enqueue(port[1], msg)

    def enqueue(self, queue, val):
        if len(queue) >= 1:
            queue.pop()
        queue.append(val)

    def sendOutputs(self):
        for port in self.outPortTupleVector:
            applicationQueue = port[0]
            if len(applicationQueue) != 0:
                msg = applicationQueue[0]
                applicationQueue.pop()
                self.enqueue(port[1], msg)

        for port in self.outPortTupleVector:
            infrastructureQueue = port[1]
            if len(infrastructureQueue) != 0:
                msg = infrastructureQueue[0]
                infrastructureQueue.pop()
                (port[2])(msg)

File: resources/expected/ros2/isolette_inverted_topics/strict/src/isolette_cpp_pkg/src/base_code/temperature_sensor_cpi_thermostat_base_src.cpp
#include "isolette_cpp_pkg/base_headers/temperature_sensor_cpi_thermostat_base_src.hpp"

//========================================================
// Re-running Codegen will overwrite changes to this file
//========================================================

temperature_sensor_cpi_thermostat_base::temperature_sensor_cpi_thermostat_base() : Node("temperature_sensor_cpi_thermostat")
{
    cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
    subscription_options_.callback_group = cb_group_;

    // Setting up connections
    temperature_sensor_cpi_thermostat_air_subscription_ = this->create_subscription<isolette_cpp_pkg_interfaces::msg::PhysicalTempimpl>(
        "temperature_sensor_cpi_thermostat_air",
        1,
        std::bind(&temperature_sensor_cpi_thermostat_base::accept_air, this, std::placeholders::_1), subscription_options_);

    temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_ = this->create_publisher<isolette_cpp_pkg_interfaces::msg::TempWstatusimpl>(
        "temperature_sensor_cpi_thermostat_current_tempWstatus",
        1);

    // timeTriggeredCaller callback timer
    periodTimer_ = this->create_wall_timer(std::chrono::milliseconds(1000),
        std::bind(&temperature_sensor_cpi_thermostat_base::timeTriggeredCaller, this), cb_group_);

    // Used by receiveInputs
    inDataPortTupleVector = {
        {&infrastructureIn_air, &applicationIn_air}
    };

    // Used by receiveInputs
    inEventPortTupleVector = {
        {&infrastructureIn_air, &applicationIn_air}
    };

    // Used by sendOutputs
    outPortTupleVector = {
        {&applicationOut_current_tempWstatus, &infrastructureOut_current_tempWstatus, &temperature_sensor_cpi_thermostat_base::sendOut_current_tempWstatus}
    };

}

void temperature_sensor_cpi_thermostat_base::init_air(isolette_cpp_pkg_interfaces::msg::PhysicalTempimpl val) {
    enqueue(infrastructureIn_air, val);
}

//=================================================
//  C o m m u n i c a t i o n
//=================================================

void temperature_sensor_cpi_thermostat_base::accept_air(isolette_cpp_pkg_interfaces::msg::PhysicalTempimpl msg)
{
    enqueue(infrastructureIn_air, msg);
}

isolette_cpp_pkg_interfaces::msg::PhysicalTempimpl temperature_sensor_cpi_thermostat_base::get_air() {
    MsgType msg = applicationIn_air.front();
    return std::get<isolette_cpp_pkg_interfaces::msg::PhysicalTempimpl>(msg);
}

void temperature_sensor_cpi_thermostat_base::sendOut_current_tempWstatus(MsgType msg)
{
    if (auto typedMsg = std::get_if<isolette_cpp_pkg_interfaces::msg::TempWstatusimpl>(&msg)) {
        temperature_sensor_cpi_thermostat_current_tempWstatus_publisher_->publish(*typedMsg);
    } else {
        PRINT_ERROR("Sending out wrong type of variable on port current_tempWstatus.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.");
    }
}

void temperature_sensor_cpi_thermostat_base::put_current_tempWstatus(isolette_cpp_pkg_interfaces::msg::TempWstatusimpl msg)
{
    enqueue(applicationOut_current_tempWstatus, msg);
}

void temperature_sensor_cpi_thermostat_base::timeTriggeredCaller() {
    receiveInputs();
    timeTriggered();
    sendOutputs();
}

void temperature_sensor_cpi_thermostat_base::receiveInputs() {
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*> port : inDataPortTupleVector) {
        auto infrastructureQueue = std::get<0>(port);
        if (!infrastructureQueue->empty()) {
            auto msg = infrastructureQueue->front();
            enqueue(*std::get<1>(port), msg);
        }
    }
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*> port : inEventPortTupleVector) {
        auto infrastructureQueue = std::get<0>(port);
        if (!infrastructureQueue->empty()) {
            auto msg = infrastructureQueue->front();
            infrastructureQueue->pop();
            enqueue(*std::get<1>(port), msg);
        }
    }
}

void temperature_sensor_cpi_thermostat_base::enqueue(std::queue<MsgType>& queue, MsgType val) {
    if (queue.size() >= 1) {
        queue.pop();
    }
    queue.push(val);
}

void temperature_sensor_cpi_thermostat_base::sendOutputs() {
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (temperature_sensor_cpi_thermostat_base::*)(MsgType)> port : outPortTupleVector) {
        auto applicationQueue = std::get<0>(port);
        if (applicationQueue->size() != 0) {
            auto msg = applicationQueue->front();
            applicationQueue->pop();
            enqueue(*std::get<1>(port), msg);
        }
    }

    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (temperature_sensor_cpi_thermostat_base::*)(MsgType)> port : outPortTupleVector) {
        auto infrastructureQueue = std::get<1>(port);
        if (infrastructureQueue->size() != 0) {
            auto msg = infrastructureQueue->front();
            infrastructureQueue->pop();
            (this->*std::get<2>(port))(msg);
        }
    }
}
